chainer/functions/loss/discriminative_loss.py

```diff
@@ -0,0 +1,350 @@
+
+            _dist_term_sample = F.sum(
+                F.maximum(module.asarray(0.0, means.dtype), margin - nrm) ** 2)
+            _dist_term_sample /= nobj * (nobj - 1)
+            dist_term += _dist_term_sample
+
+        dist_term /= bs
+
+        return dist_term
+
+    def _regularization_term(self, means, n_objects):
+        """Function to calculate regularization term
+
+        Args:
+            means (nd-array): Instance means
+            n_objects (int): Instance count in current input
+
+        Returns:
+            float : distance loss
+        """
+
+        bs, n_instances, n_filters = means.shape
+        m = F.cast(means, means.dtype)
+
+        reg_term = 0.0
+        for i in range(bs):
+            reg_term += F.mean(self.norm(m[i, : n_objects[i], :], 1))
+        reg_term /= bs
+
+        return reg_term
+
+    def _means(self, pred, gt, n_objects, max_n_objects, gt_idx):
+        """Function to calculate cluster means
+
+        Args:
+            pred (nd-array): Prediction output
+            gt (nd-array): Ground truth output
+            n_objects (int): Instance number in current input
+            max_n_objects (int): Maximum possible instance number
+            gt_idx (tuple / nd-array): Indexes of ground truth instances
+
+        Returns:
+            tuple : distance loss
+        """
+
+        bs = pred.shape[0]
+        n_filters = pred.shape[1]
+        n_instances = gt.shape[1]
+        n_loc = pred.shape[2]
+
+        # Expand prediction and broadcast instances to 3rd axis
+        p = F.reshape(pred, (bs, n_filters, 1, n_loc))
+        p = F.broadcast_to(p, (bs, n_filters, n_instances, n_loc))
+
+        # Expand ground truth to match the size but do not broadcast
+        g = F.reshape(gt, (bs, 1, n_instances, n_loc))
+
+        p = p * F.cast(g, p.dtype)
+        module = cuda.get_array_module(p)
+
+        means = []
+        for i in range(bs):
+
+            p_item = p[i, :, gt_idx[i]]
+            g_item = g[i, :, gt_idx[i]]
+
+            p_sum = F.sum(p_item, axis=2)
+            g_sum = F.cast(F.sum(g_item, axis=2), p_sum.dtype)
+            _mean_sample = p_sum / g_sum
+
+            n_fill_objects = max_n_objects - n_objects[i]
+
+            if n_fill_objects != 0:
+                _fill_sample = module.zeros(
+                    (module.asnumpy(n_fill_objects), n_filters),
+                    dtype=_mean_sample.dtype)
+                _mean_sample = F.concat((_mean_sample, _fill_sample), axis=0)
+
+            means.append(_mean_sample)
+
+        means = F.stack(means)
+        return means
+
+    def _prepare_inputs(self, prediction, labels):
+        """Function to preprocess inputs
+
+        Args:
+            prediction (nd-array): Prediction output
+            labels (nd-array): Ground truth output
+        Returns:
+            nd-array : Prediction output
+            nd-array : Ground truth output
+        """
+        # Reshape layers to prepare for processing
```

Comment:
I think it would be better to write this comment to the document of this function.